mesonbuild/interpreterbase/interpreterbase.py

```diff
@@ -169,25 +168,21 @@ def evaluate_codeblock(self, node: mparser.CodeBlockNode, start: int = 0, end: T
         if node is None:
             return
         if not isinstance(node, mparser.CodeBlockNode):
-            e = InvalidCode('Tried to execute a non-codeblock. Possibly a bug in the parser.')
-            e.lineno = node.lineno
-            e.colno = node.colno
-            raise e
+            raise InvalidCode.from_node('Tried to execute a non-codeblock. Possibly a bug in the parser.', node=node)
         statements = node.lines[start:end]
         i = 0
         while i < len(statements):
             cur = statements[i]
+            self.current_lineno = cur.lineno
             try:
-                self.current_lineno = cur.lineno
                 self.evaluate_statement(cur)
+            except MesonException as e:
+                # Assume that if file is set, then the information in the exception is correct
+                if e.file is None:
+                    e.set_from_node(self.current_node)
+                raise
             except Exception as e:
-                if getattr(e, 'lineno', None) is None:
-                    # We are doing the equivalent to setattr here and mypy does not like it
-                    # NOTE: self.current_node is continually updated during processing
```

Comment:
We should keep this note.

Reply:
I added the comment back in the commit that removes this line.

Then I finally came up with what I thought is a good way to just solve this problem by getting rid of `cur` alltogether, there is now only `self.current_node`. Let me konw what you think